src/model.ts

```diff
@@ -0,0 +1,191 @@
+import { IDisposable } from '@lumino/disposable';
+import { ISignal, Signal } from '@lumino/signaling';
+
+import { ChatService } from './services';
+
+export interface IChatModel extends IDisposable {
+  /**
+   * The chat model ID.
+   */
+  id: string;
+
+  /**
+   * The signal emitted when a new message is received.
+   */
+  get incomingMessage(): ISignal<IChatModel, ChatService.IMessage>;
+
+  /**
+   * The signal emitted when a message is updated.
+   */
+  get messageUpdated(): ISignal<IChatModel, ChatService.IMessage>;
+
+  /**
+   * The signal emitted when a message is updated.
+   */
+  get messageDeleted(): ISignal<IChatModel, ChatService.IMessage>;
+
+  /**
+   * Send a message, to be defined depending on the chosen technology.
+   * Default to no-op.
+   *
+   * @param message - the message to send.
+   * @returns whether the message has been sent or not, or nothing if not needed.
+   */
+  sendMessage(
```

Comment:
Just a naming question. I'm wondering if `addMessage` would be better.

In my mind, in this place we're changing the chat model to add a new message into it. Whether it is "sent" is entirely to the chatmodel implementation detail.

```suggestion
  addMessage(
```

Reply:
When looking into implementing a CRDT-based chat, we'd add a message to the data structure. The sending of the changes would not be handled by the chat implementation.

Reply:
So I wouldn't make the base ChatModel depend on `ChatService`, it seems to me `ChatService` would be an implementation detail of the websocket chat implementation?

Reply:
> it seems to me ChatService would be an implementation detail of the websocket chat implementation?

It is true for part of it, but some of the types/interfaces are more generics and used by the view.
I will split it, to import only the required types.